### PR TITLE
fix douyutv issue #637

### PR DESCRIPTION
**Pull request by fozzysec**
fix issue #637
Thanks to @loboque 's report and help from @yan12125 @brlala
New algorithm from @spacemeowx2 and its python implement by @debugzxcv

### CODE REVIEW

**Comment by fozzysec** [comment]
The original python algorithm has minor bug, and can be fixed by
```
--- src/streamlink/plugins/douyutv_blackbox.py	2017-03-04 10:54:39.000000000 +0800
+++ origin_douyutv_blackbox.py	2017-03-04 11:00:13.000000000 +0800
@@ -7,7 +7,7 @@
     # s = '8101252B10B1008684E2B988C3C72A33F832Ba2053899224e8a92974c729dceed1cc99b3d828224809284'
     hashstr = md5(s)
     mid = [int(hashstr[i:i + 2], 16) for i in range(0, len(hashstr), 2)]
-    F_func_173e8124cdbdc90d([c for c in s], mid)
+    F_func_173e8124cdbdc90d([ord(c) for c in s], mid)
     return ''.join('{:02x}'.format(x) for x in mid)
 
 
@@ -201,7 +201,7 @@
     output = [0] * ((len(input) >> 2) + 1)
     length8 = len(input) * 8
     for i in range(0, length8, 8):
-        output[i >> 5] |= ((input[int(i / 8)]) & 0xFF) << i % 32
+        output[i >> 5] |= (ord(input[int(i / 8)]) & 0xFF) << i % 32
 
     return output
 
```

File: src/streamlink/plugins/douyutv_blackbox.py [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
...0xdf, 0x5c, 0x39, 0xeb, 0xa0, 0xb0, 0xb2, 0x53, 0x2f, 0xe7, 0x7c, 0xf3, 0x3e, 0x4a, 0x84, 0xd6, 0xc4, 0x5e, 0xab, 0xf1, 0x5f, 0xaf, 0x95, 0xd4, 0x6a, 0xfd, 0x10, 0xfc, 0x9f, 0xc0, 0x7a, 0x4, 0xfa, 0xa5, 0x1f, 0x1d, 0xc5, 0x1a, 0x23, 0x4e, 0x9d, 0xd5, 0x70, 0x90, 0x85, 0xe2, 0x76, 0x43, 0x20, 0x2c, 0x0, 0x45, 0xd1, 0x13, 0x28, 0xdd, 0xde, 0x33, 0xd2, 0xb9, 0xf7, 0x87, 0xf5, 0xbb, 0x9e, 0x65, 0xa4, 0xae, 0x93, 0xa1, 0x98, 0x2, 0xf, 0x58, 0x7, 0xa8, 0xdb, 0x2b, 0xc, 0x79, 0x88, 0x6, 0x62, 0x97, 0x42, 0x91, 0x9b, 0x82, 0x22, 0x5d, 0xcc, 0x8e, 0x75, 0x5a, 0xc8, 0x83, 0xca, 0xef, 0x8d, 0x31, 0x99, 0x8c, 0x18, 0x19, 0x74, 0xe1, 0x35, 0xcd, 0x24, 0x69, 0xda, 0x48, 0x6e, 0xc2, 0x15, 0xc7, 0x11, 0x21, 0x63, 0x38, 0x46, 0x5, 0x71, 0xe, 0x59, 0x36, 0x94, 0x81, 0xb8, 0x7e, 0x89, 0xbe, 0x3a, 0xee, 0xa3, 0x7b, 0x1e, 0xe0, 0xe8, 0x41, 0x66, 0xd9, 0x51, 0x14, 0x67, 0xb5, 0x6c, 0x3c, 0x34, 0x3b, 0xba, 0xc9, 0x4b, 0xa7, 0x49, 0xaa, 0xf9, 0x37, 0x30, 0x2a, 0x72, 0xe4, 0xa9, 0x96, 0xcb, 0x27, 0xce, 0xc3, 0x55, 0xad, 0x4d, 0x32, 0x54, 0xfb, 0xd7, 0xa, 0x61, 0x80, 0x77, 0x73, 0xd8, 0x8a, 0xb6, 0xe3, 0xfe, 0xe9, 0x1b, 0x29, 0xec, 0x56, 0x12, 0x60, 0xf0, 0xa6, 0xf8, 0xed, 0x7f, 0x44, 0x40, 0x1c, 0x6f, 0x3]
```
Comment (gravyboat) [inline src/streamlink/plugins/douyutv_blackbox.py]: Looks like this table is causing some issues with Python 2.7 on Travis builds.

File: src/streamlink/plugins/douyutv.py [hunk 2]
```diff
@@ -117,7 +123,9 @@ def _get_streams(self):
 
         ts = int(time.time() / 60)
         did = uuid.uuid4().hex.upper()
-        sign = hashlib.md5(("{0}{1}{2}{3}".format(channel, did, LAPI_SECRET, ts)).encode("utf-8")).hexdigest()
+
+        #use new API key and modified MD5 algorithm
+        sign = stupidMD5(("{0}{1}{2}{3}".format(channel, did, LAPI_SECRET, ts)).encode("utf-8"))
```
Comment (steven7851) [inline src/streamlink/plugins/douyutv.py]: should be ```sign = stupidMD5("{0}{1}{2}{3}".format(channel, did, LAPI_SECRET, ts))```

File: src/streamlink/plugins/douyutv_blackbox.py (hunk 1 above)
Comment (fozzysec) [inline src/streamlink/plugins/douyutv_blackbox.py]: I am looking into the algorithm with python 2, I only tested with python3.

Comment (spacemeowx2) [inline src/streamlink/plugins/douyutv_blackbox.py]: Maybe you need to add `# coding: xxx`.

**Comment by fozzysec** [comment]
I encoded it with utf8 which caused python3 error, by modifying `stupidMD5` it works but python2 fails of bit operator of different types. Now got it fixed and work both in python2 and python3

File: src/streamlink/plugins/douyutv_blackbox.py [hunk 3]
```diff
@@ -0,0 +1,986 @@
...
+        j = lens >> 3
+        F_func_5601962242a657f3(s, i, locTable)
+        i = i + 8
+        j = j - 1
+        while j != 0:
+            F_func_5601962242a657f3(s, i, locTable)
+            i = i + 8
+            j = j - 1
+
+    # 剩余的和key异或
```
Comment (beardypig) [inline src/streamlink/plugins/douyutv_blackbox.py]: Did you try removing this comment or setting the encoding?

File: src/streamlink/plugins/douyutv_blackbox.py (hunk 1 above)
Comment (beardypig) [inline src/streamlink/plugins/douyutv_blackbox.py]: What @spacemeowx2 suggests fixes the Python 2.7 tests. Make `#coding: utf8` the first line of the file.

File: src/streamlink/plugins/douyutv_blackbox.py (hunk 3 above)
Comment (fozzysec) [inline src/streamlink/plugins/douyutv_blackbox.py]: I will translate Chinese comments to English, sorry for the late replying

**Comment by beardypig** [comment]
Thanks for the fix @fozzysec :)

**Comment by gravyboat** [comment]
https://github.com/streamlink/streamlink/pull/671 is an extension on this to clean things up.

**Pull request merged**